### ISSUE TITLE
HTOP reporting 16.0Z as memory usage in alpine 3.14 in LXD / Ubuntu 18.04

### ISSUE THREAD

**Issue by lepe** [htop-dev/htop#863]
When using Alpine Linux 3.14 (amd64) as container, memory inside htop is always reported as "16.0Z" while in previous versions it shows the correct value. Values inside top command seems correct.

How to reproduce:

```
lxc launch images:alpine/3.14 ram-test
lxc exec ram-test apk add htop
lxc exec ram-test htop
```

What to expect:

In memory "bar", it should display the correct value, for example: 8M

@aither64 :
> I have just been trying to figure out a similar issue... I think it's caused by htop reading /proc/spl/kstat/zfs/arcstats and then substracting the size of ZFS ARC from the used memory read from /proc/meminfo... you can see it in the strace output and e.g. here in the htop code: https://github.com/htop-dev/htop/blob/main/linux/Platform.c#L342

@stgraber :
> Can you try mount -t tmpfs tmpfs /proc/spl and see if the problem goes away then?
> Hmm, okay, then I think we need a bug report against htop as it seems like it's being slightly too smart for its own good on this one :)

The issue is resolved after executing such command. Here is the full report:
https://github.com/lxc/lxcfs/issues/489

**Comment by BenBE** [comment]
Which version of `htop` are you using (<kbd>F1</kbd> inside `htop` should tell)?

What does `/proc/spl/kstat/zfs/arcstats` look like?

Can you pin down an old version that this issue did not yet appear in?

**Comment by aither64** [comment]
To expand on my comment quoted above... LXC/LXD containers use LXCFS to override certain files in `/proc` and `/sys` to show values specific to containers. For example, `/proc/meminfo` is generated from the container's memory cgroup. This is to ensure that tools reading `/proc/meminfo` report values relevant to the container it is run in and not the host system.

As I understand it, the issue is that htop unconditionally substracts ZFS ARC size from used memory calculated from `/proc/meminfo`. This won't work with LXCFS, because ZFS ARC is not accounted by the memory cgroup, i.e. it's not included in values in its `/proc/meminfo`. Since the ARC is often larger than the memory used by containers, it results in an overflow.

It's been introduced by dfa40ad0ebdb40a859ea16a264e775ce766bb037 in 3.0.3. 3.0.2 reports correct memory usage to me. I'm not sure how to best solve it unfortunately.

FWIW, on my systems, we've run `chmod o-r /proc/spl/kstat/zfs/arcstats` on the host, which makes it inaccessible to unprivileged containers and htop will ignore it.

**Comment by fasterit** [comment]
When we walk processes we should use `/proc/1/cgroup` to detect whether `htop` runs under these, set a flag and in turn omit substracting the ARC cache.

**Comment by BenBE** [comment]
Or skip substracting the ARC when it's larger than the available memory.

From inside a container you can't necessarily detect that you are inside one.

**Comment by fasterit** [comment]
@BenBE The subtract only when it's larger than available memory is a bad heuristic as it will misfire when people have lots of RAM or little ARC caches.

Currently you can identify for sure that you are in one of the affected container runtimes. What you are thinking of is more full VMs that - for security analysis reasons - hide that they are in fact a VM. These would not see the host ARC cache anyways. So we are only dealing with the half-baked-fake-memory-reporting types ... container runtimes.

**Comment by BenBE** [comment]
> @BenBE The subtract only when it's larger than available memory is a bad heuristic as it will misfire when people have lots of RAM or little ARC caches.
> 
> Currently you can identify for sure that you are in one of the affected container runtimes. What you are thinking of is more full VMs that - for security analysis reasons - hide that they are in fact a VM. These would not see the host ARC cache anyways. So we are only dealing with the half-baked-fake-memory-reporting types ... container runtimes.

@fasterit Okay, if you document this assumption properly about being able to identify being inside a container, fine by me.

And regarding the heuristic: I actually only just meant the case `ARC > totalMem` / `ARC > usedMem`, i.e. when we know this will underflow and thus report incorrect values. At least as a fallback this might be worth having.

**Comment by lepe** [comment]
> Which version of `htop` are you using (F1 inside `htop` should tell)?
3.0.5

> Can you pin down an old version that this issue did not yet appear in?
2.2.0 (from Alpine 3.10 which was used before, worked fine).

> What does `/proc/spl/kstat/zfs/arcstats` look like?

```
root@host:~# cat /proc/spl/kstat/zfs/arcstats

13 1 0x01 96 4608 10059903377 13656364245791935
name                            type data
hits                            4    273231292
misses                          4    29874571
demand_data_hits                4    777
demand_data_misses              4    0
demand_metadata_hits            4    270820957
demand_metadata_misses          4    28206750
prefetch_data_hits              4    0
prefetch_data_misses            4    0
prefetch_metadata_hits          4    2409558
prefetch_metadata_misses        4    1667821
mru_hits                        4    70287784
mru_ghost_hits                  4    5196704
mfu_hits                        4    202404552
mfu_ghost_hits                  4    980291
deleted                         4    8293125
mutex_miss                      4    2217871
access_skip                     4    15
evict_skip                      4    1157033197
evict_not_enough                4    17522707
evict_l2_cached                 4    0
evict_l2_eligible               4    369534947328
evict_l2_ineligible             4    19688478720
evict_l2_skip                   4    0
hash_elements                   4    75540
hash_elements_max               4    100207
hash_collisions                 4    506319
hash_chains                     4    2962
hash_chain_max                  4    4
p                               4    86332785
c                               4    585132400
c_min                           4    260331520
c_max                           4    4165304320
size                            4    574996288
compressed_size                 4    29422592
uncompressed_size               4    170074112
overhead_size                   4    160274432
hdr_size                        4    25538256
data_size                       4    0
metadata_size                   4    189697024
dbuf_size                       4    76556016
dnode_size                      4    205660352
bonus_size                      4    77544640
anon_size                       4    34816
anon_evictable_data             4    0
anon_evictable_metadata         4    0
mru_size                        4    52371968
mru_evictable_data              4    0
mru_evictable_metadata          4    498176
mru_ghost_size                  4    156576256
mru_ghost_evictable_data        4    59343872
mru_ghost_evictable_metadata    4    97232384
mfu_size                        4    137290240
mfu_evictable_data              4    0
mfu_evictable_metadata          4    1228800
mfu_ghost_size                  4    219042816
mfu_ghost_evictable_data        4    0
mfu_ghost_evictable_metadata    4    219042816
l2_hits                         4    0
l2_misses                       4    0
l2_feeds                        4    0
l2_rw_clash                     4    0
l2_read_bytes                   4    0
l2_write_bytes                  4    0
l2_writes_sent                  4    0
l2_writes_done                  4    0
l2_writes_error                 4    0
l2_writes_lock_retry            4    0
l2_evict_lock_retry             4    0
l2_evict_reading                4    0
l2_evict_l1cached               4    0
l2_free_on_write                4    0
l2_abort_lowmem                 4    0
l2_cksum_bad                    4    0
l2_io_error                     4    0
l2_size                         4    0
l2_asize                        4    0
l2_hdr_size                     4    0
memory_throttle_count           4    0
memory_direct_count             4    10835
memory_indirect_count           4    132201
memory_all_bytes                4    8330608640
memory_free_bytes               4    2282733568
memory_available_bytes          3    2152570880
arc_no_grow                     4    0
arc_tempreserve                 4    0
arc_loaned_bytes                4    0
arc_prune                       4    63
arc_meta_used                   4    574996288
arc_meta_limit                  4    3123978240
arc_dnode_limit                 4    312397824
arc_meta_max                    4    925683864
arc_meta_min                    4    16777216
sync_wait_for_async             4    2163
demand_hit_predictive_prefetch  4    477506
arc_need_free                   4    0
arc_sys_free                    4    130165760
```

**Comment by d0x7** [comment]
Pretty sure I have the same issue. I'll attach my arcstats, but I guess won't do much of a difference. 

I'm very certain it's the same issue, that exists since at least mid 2016. Has been discussed over on [lxd](https://github.com/lxc/lxd/issues/2108), then on [lxcfs](https://github.com/lxc/lxcfs/issues/120), on [htop ](https://github.com/hishamhm/htop/issues/599) and now on htop-dev again.

I'm on Proxmox 7.0-13 using LXC 4.0.5 and LXCFS 4.0.8. The containers have htop 3.0.5 and are mostly on debian.
<details>
  <summary>cat /proc/spl/kstat/zfs/arcstats</summary>

![MobaXterm_dPGdl4HcCP](https://user-images.githubusercontent.com/17951467/140818431-16d92fd7-6a1b-40fb-be69-6a82a56e4df2.jpg)
  ```bash
root@pve-lxchtopdev:~#  cat /proc/spl/kstat/zfs/arcstats
13 1 0x01 116 31552 5112315670 97695212546990
name                            type data
hits                            4    29123033
misses                          4    934424
demand_data_hits                4    12211095
demand_data_misses              4    54006
demand_metadata_hits            4    16876598
demand_metadata_misses          4    45008
prefetch_data_hits              4    11427
prefetch_data_misses            4    824374
prefetch_metadata_hits          4    23913
prefetch_metadata_misses        4    11036
mru_hits                        4    8217941
mru_ghost_hits                  4    12329
mfu_hits                        4    20883543
mfu_ghost_hits                  4    28595
deleted                         4    1889431
mutex_miss                      4    5
access_skip                     4    15
evict_skip                      4    37863
evict_not_enough                4    697
evict_l2_cached                 4    0
evict_l2_eligible               4    124457493504
evict_l2_ineligible             4    7945654784
evict_l2_skip                   4    0
hash_elements                   4    560991
hash_elements_max               4    675399
hash_collisions                 4    378210
hash_chains                     4    34206
hash_chain_max                  4    4
p                               4    2010166784
c                               4    16792125440
c_min                           4    1049507840
c_max                           4    16792125440
size                            4    16798654560
compressed_size                 4    12549683200
uncompressed_size               4    14083123712
overhead_size                   4    1473991168
hdr_size                        4    188853072
data_size                       4    12542658560
metadata_size                   4    1481015808
dbuf_size                       4    604265088
dnode_size                      4    1493277072
bonus_size                      4    484171520
anon_size                       4    1536
anon_evictable_data             4    0
anon_evictable_metadata         4    0
mru_size                        4    1428518400
mru_evictable_data              4    4470784
mru_evictable_metadata          4    92160
mru_ghost_size                  4    15256016896
mru_ghost_evictable_data        4    9986900480
mru_ghost_evictable_metadata    4    5269116416
mfu_size                        4    12595154432
mfu_evictable_data              4    11995107328
mfu_evictable_metadata          4    89271808
mfu_ghost_size                  4    1532645376
mfu_ghost_evictable_data        4    1007812608
mfu_ghost_evictable_metadata    4    524832768
l2_hits                         4    0
l2_misses                       4    0
l2_feeds                        4    0
l2_rw_clash                     4    0
l2_read_bytes                   4    0
l2_write_bytes                  4    0
l2_writes_sent                  4    0
l2_writes_done                  4    0
l2_writes_error                 4    0
l2_writes_lock_retry            4    0
l2_evict_lock_retry             4    0
l2_evict_reading                4    0
l2_evict_l1cached               4    0
l2_free_on_write                4    0
l2_abort_lowmem                 4    0
l2_cksum_bad                    4    0
l2_io_error                     4    0
l2_size                         4    0
l2_asize                        4    0
l2_hdr_size                     4    0
l2_log_blk_writes               4    0
l2_log_blk_avg_asize            4    0
l2_log_blk_asize                4    0
l2_log_blk_count                4    0
l2_data_to_meta_ratio           4    0
l2_rebuild_success              4    0
l2_rebuild_unsupported          4    0
l2_rebuild_io_errors            4    0
l2_rebuild_dh_errors            4    0
l2_rebuild_cksum_lb_errors      4    0
l2_rebuild_lowmem               4    0
l2_rebuild_size                 4    0
l2_rebuild_asize                4    0
l2_rebuild_bufs                 4    0
l2_rebuild_bufs_precached       4    0
l2_rebuild_log_blks             4    0
memory_throttle_count           4    0
memory_direct_count             4    0
memory_indirect_count           4    730
memory_all_bytes                4    33584250880
memory_free_bytes               4    9762902016
memory_available_bytes          3    8537491456
arc_no_grow                     4    0
arc_tempreserve                 4    0
arc_loaned_bytes                4    0
arc_prune                       4    75971
arc_meta_used                   4    4251582560
arc_meta_limit                  4    12594094080
arc_dnode_limit                 4    1259409408
arc_meta_max                    4    4264723040
arc_meta_min                    4    16777216
async_upgrade_sync              4    160617
demand_hit_predictive_prefetch  4    661984
demand_hit_prescient_prefetch   4    0
arc_need_free                   4    0
arc_sys_free                    4    1225410560
arc_raw_size                    4    0
cached_only_in_progress         4    0
abd_chunk_waste_size            4    4413440
  ```
</details>

**Comment by IndrekHaav** [comment]
I've also experienced this in LXC containers running on Proxmox. Debian 11 containers with htop 3.0.5 exhibit the problem, Debian 10 and Ubuntu 20.04 containers with htop 2.2.0 don't.

The workaround mentioned by @aither64 solves the issue for me:

> FWIW, on my systems, we've run `chmod o-r /proc/spl/kstat/zfs/arcstats` on the host, which makes it inaccessible to unprivileged containers and htop will ignore it.

**Comment by fasterit** [comment]
Related as also a LXD-hides-stuff issue ... #964 / #965

**Comment by er-azh** [comment]
Actually, the CPU issue was because lxcfs was reporting `DT_UNKNOWN` as `d_type` for all CPU entries in `/sys/devices/system/cpu/` and `LinuxProcessList_updateCPUcount` couldn't find any CPUs, leaving `cpuData` uninitialized. 
 
But I think htop should still initialize `cpuData` even when no CPU entry is found.

**Comment by fasterit** [comment]
Neither parsing `/proc/1/cgroup` nor `/proc/1/sched` survived testing. So #967 is using `/proc/1/mounts` which seems sufficient for our use case and is easier to check than `mountinfo`. Also available in kernel 2.4 vs 2.6.

Please test the (draft) PR on docker and lxc.

**Comment by NewEraCracker** [comment]
> I have the same Issue with LXC and debian 11.

@KxmischerLasse Looks like `mount -t tmpfs tmpfs /proc/spl/kstat/zfs` is a valid workaround (run inside the container).

So I have added to the container's crontab: `@reboot mount -t tmpfs tmpfs /proc/spl/kstat/zfs` while waiting for the Debian folks to release the updated version with the fix.

My two cents, no warranties express or implied.
